man/man1/stoo.1
.TH STOO 1NEMO "8 February 2012" 

.SH "NAME"
stoo \- convert snapshot to orbit

.SH "SYNOPSIS"
.PP
\fBstoo in=\fPsnapshot \fBout=\fPorbit [parameter=value]

.SH "DESCRIPTION"
\fBstoo\fP selects one or more bodies from a snapshot file
and converts it to orbit format. Orbits are
stored sequentially, the Key field in the snapshot is used to
identify the particle making this orbit. If no Key field is present,
the first body has key 0.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fIin-file\fP
input file, in \fIsnapshot(5NEMO)\fP format. If in a pipe, the
number of snapshots to be converted to orbits needs to be
given via \fBnsteps=\fP, see below. [no default]
.TP
\fBout=\fIout-file\fP
output file, will be in \fIorbit(5NEMO)\fP format [no default]
.TP
\fBibody=\fInumber\fP
ranking \fInumber\fP of the star in the snapshot file,
normally the Key field.  Multiple stars can be selected here,
e.g. 0:1000:10 in the usual \fInemoinp(3NEMO)\fP notation.
However, all orbits need to be allocated in memory. Use a negative
number to process all bodies.
[default: \fB-1\fP].
.TP
\fBnsteps=\fInumber\fP
\fInumber\fP of timesteps to be allocated. Use 0 to allow the program
to scan the input snapshot. However, if the input is from a pipe,
\fBnsteps\fP needs to be given the (maximum) number of steps allowed.

[default: \fB0\fP].

.SH "EXAMPLES"
Here is an example of producing a large snapshot, in single precision, with
\fIgyrfalcon(1NEMO)\fP. Since currently \fIstoo(1NEMO)\fP cannot read such
single precision Positon/Velocity yet, they need to be converted via
\fIsnapcopy(1NEMO)\fP. Some timings are given.  The input snapshot is 430MB,
the output snapshot and orbit file are 860MB.
.nf

   mkplummer p10k 10000
   gyrfalcON p10k p10k.out tstop=100 step=0.1 kmax=8 eps=0.1 give=mxvap > p10k.log

   /usr/bin/time snapcopy p10k.out tmp.out
   2.52user 0.48system 0:03.02elapsed 99%CPU


   /usr/bin/time stoo tmp.out tmp.orb -1
   1.60user 1.04system 0:02.65elapsed 99%CPU

   /usr/bin/time orbstat tmp.orb > tmp.orb.log
   0.49user 0.23system 0:00.73elapsed 99%CPU

.fi

Data-wide this "Particle-Attribute-Time" dataset  is 10000 x 11 x 1000 in this case,
but at various stages dimensions are shuffled around rather haphazardly, and
does not scale well for large numbers of particles and/or times.


.SH "SEE ALSO"
snapshot(5NEMO), orbit(5NEMO), orblist(1NEMO), otos(1NEMO), orboom(1NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +2.5i
~/src/orbit/misc   	stoo.c
.fi

.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
27-jul-87	V1.0 created	PJT
8-jun-88	V2.1 new filestruct	PJT
23-jun-91	V2.4 fixed bugs, upgraded doc for NEMO2 	PJT
24-jul-2013	V3.0 added support for multiple orbits and use Key	PJT
11-nov-2015	V3.1 efficient version without rewinding file	PJT
6-feb-2022	V3.3 process all particles by default, init I1	PJT
.fi


src/orbit/misc/stoo.c
/*
 *  STOO:  convert a binary snapshot file to a binary orbit
 *
 *	27-jul-87	V1.0 original version	Peter Teuben	
 *	28-jul-87	V2.0 new orbit(5)	PJT
 *	8-jun-88	V2.1 new filestruct	PJT
 *	30may-90	V2.2 fixed Masso()	PJT 
 *	21-nov-90	V2.3 new Nemo		PJT
 *	12-jun-91	V2.4 time->tsnap fixes bug, also process history PJT
 *	 7-mar-92	     make gcc happy
 *	25-may-92	V2.5 need <potential.h> now
 *	18-apr-95       V2.6 
 *      25-jul-2013     V3.0 allow multiple orbits output;
 *      11-nov-2015     V3.1 now the efficient way
 *      19-dec-2019     V3.2 deal with ACC/PHI
 *       6-feb-2022     V3.3 initialize I1, properly implement ibody=-1
 *       8-feb-2022     V3.4 stopped massive memory leak, add scan_snap() for files
 *                           use mdarray for dynamic snapshot
 *
 *  @todo    deal with no Potential/Accelleration
 */

#include <stdinc.h>
#include <unistd.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <mdarray.h>

#include <snapshot/snapshot.h>
#include <potential.h>
#include <orbit.h>

string defv[] = {
  "in=???\n			snapshot input file name",
  "out=???\n			orbit output file name",
  "ibody=-1\n			which particles to take (-1=all, 0=first)",
  "nsteps=0\n	                max orbit length allocated in case of pipe",
  "VERSION=3.4b\n		9-feb-2022 PJT",
  NULL,
};

string usage="convert a snapshot file to an orbit";
string cvsid="$Id$";

string iname, oname;		/* file names */
stream instr,outstr;		/* file pointers */
string headline;

int    nobj;                    /* number of bodies per snapshot */
int    norb;                    /* number of bodies to extract for orbits */
int    nsteps;                  /* number of snapshots to use for each orbit */
real   tsnap;			/* time of snapshot */

#define USE_MDARRAY

#if defined(USE_MDARRAY)
  int MOBJ = 0;
  real *mass;
  real ***phase;
  real *phi;
  real **acc;
  int  *key;
  int  *isel;
#else
  #define MOBJ 	1000000
  real   mass[MOBJ];
  real   phase[MOBJ][2][NDIM];
  real   phi[MOBJ];
  real   acc[MOBJ][NDIM];
  int    key[MOBJ];
  int    isel[MOBJ];              /* select for output orbit? */
#endif


bool   Qtime, Qmass, Qkey;

int    ibody;	
orbitptr *optr;			/* pointer to orbit pointers */

int read_snap();
int scan_snap();
void alloc_snap(int);

bool ispipe(stream instr)
{
  off_t try = lseek(fileno(instr), 0, SEEK_CUR);
  if (try < 0) return TRUE;
  return FALSE;
}

void nemo_main()
{
  int i,j;

  iname = getparam("in");
  oname = getparam("out");
  instr  = stropen(iname,"r");		/* read from snapshot */
  outstr = stropen(oname,"w");		/* write to orbit */
  nsteps = getiparam("nsteps");

  if (nsteps == 0)  {
    if (ispipe(instr))
      error("Need to set nsteps= for pipe files");

    nsteps = scan_snap();
    dprintf(0,"Found %d snapshots with nobj=%d\n",nsteps,nobj);
    alloc_snap(nobj);
    rewind(instr);
  } else {
    dprintf(0,"Setting orbits for %d snapshots\n",nsteps);
  }

  i = 0;				/* counter of snapshots/timesteps */
  while (read_snap()) {
      if (i==0) {			/* first time around */
	norb = nemoinpi(getparam("ibody"),isel,MOBJ);
	if (norb < 0)  error("%d: ibody=%s bad",norb,getparam("ibody"));
	if (norb == 0) error("no orbits will be output");
	
	if (isel[0] < 0) {              /* special case selecting all */
	  dprintf(1,"Selecting all %d bodies for conversion to an orbit\n", nobj);
	  norb = nobj;
	  for (j=0; j<norb; j++)
	    isel[j] = j;
	}
	
	optr=(orbitptr *) allocate(norb*sizeof(orbitptr));
	for (j=0; j<norb; j++) {
	  optr[j] = NULL;
	  allocate_orbit (&optr[j],NDIM,nsteps);
	  
	  ibody = isel[j];
	  if (ibody < 0 || ibody>=nobj)
	    error("illegal ibody=%d,  nobj=%d",ibody,nobj);
	  Masso(optr[j]) = mass[ibody];
	  I1(optr[j]) = I2(optr[j]) = I3(optr[j]) = 0.0;
	}
	dprintf(0,"Selecting %d bodies for orbits, %d..%d\n",norb,isel[0],isel[norb-1]);
  
      } else if (i>=nsteps) {
	warning("more snapshots found, stopped at %d",i);
	break;
      }
      for (j=0; j<norb; j++) {
	ibody = isel[j];
	Key(optr[j])    = key[ibody];
	Torb(optr[j],i) = tsnap;
	Xorb(optr[j],i) = phase[ibody][0][0];	
	Yorb(optr[j],i) = phase[ibody][0][1];	
	Zorb(optr[j],i) = phase[ibody][0][2];	
	Uorb(optr[j],i) = phase[ibody][1][0];	
	Vorb(optr[j],i) = phase[ibody][1][1];	
	Worb(optr[j],i) = phase[ibody][1][2];
#ifdef ORBIT_PHI
	Porb(optr[j],i) = phi[ibody];
	AXorb(optr[j],i) = acc[ibody][0];
	AYorb(optr[j],i) = acc[ibody][1];
	AZorb(optr[j],i) = acc[ibody][2];
	if (i==0) 
	  I1(optr[j]) = phi[ibody] + 0.5*(sqr(phase[ibody][1][0]) + sqr(phase[ibody][1][1]) + sqr(phase[ibody][1][2]));
#endif
      }
      i++;
      for (j=0; j<norb; j++)
	Nsteps(optr[j]) = i;			/* record actual number */
      progress(1.0,"processed snapshot %d/%d", i,nsteps);      
  }
  strclose(instr);
  dprintf(0,"Writing %d orbits\n",norb);

  if (norb > 0) {
    put_history(outstr);
    for (j=0; j<norb; j++) {
      progress(1.0,"writing orbit %d", j);
      write_orbit(outstr,optr[j]);		/* write orbit to file */
      free_orbit(optr[j]);
    }
    free(optr);
  }
  dprintf(0,"Processed %d snapshots\n",i);
  
  strclose(outstr);		        	/* close files */
}



void alloc_snap(int nbody)
{
#if defined(USE_MDARRAY)
  dprintf(1,"alloc_snap(%d)\n",nbody);
  mass = (real *) allocate_mdarray1(nbody);
  phase = (real ***) allocate_mdarray3(nbody,2,NDIM);
  phi =  (real *) allocate_mdarray1(nbody);
  acc = (real **) allocate_mdarray2(nbody,NDIM);
  key =  (int *) allocate(nbody * sizeof(int));
  isel = (int *) allocate(nbody * sizeof(int));
  MOBJ = nbody;
#else
  warning("alloc_snap: already allocated statically");
#endif  
}
  

/*
 * SCAN_SNAP:   scan to find how many snapshots there are;
 *              also ensuring they are all the same size
 */

int scan_snap()
{
  int nsnap = 0;
  int maxobj = 0;

  dprintf(0,"Scanning snapshot\n");
  for(;;) {
    get_history(instr);
    if (!get_tag_ok(instr,SnapShotTag))
      return nsnap;

     get_set(instr,SnapShotTag);
     get_set(instr, ParametersTag);
     get_data(instr, NobjTag, IntType, &nobj, 0);
     if (maxobj == 0) 
       maxobj = nobj;
     if (maxobj != nobj)
       error("Cannot process snapshots with unequal number of bodies: %d != %d", nobj,maxobj);
     get_tes(instr, ParametersTag);
     if (!get_tag_ok(instr,ParticlesTag)) {
       get_tes(instr,SnapShotTag);
       continue;
     }
     get_tes(instr,SnapShotTag);
     nsnap++;
  }
}

/*
 * READ_SNAP: read next snapshot from input stream
 *
 */

int read_snap()
{				
  int i, cs;
  static bool first = TRUE;
  static int no_mass = 0;
    
  for(;;) {  /* loop until one snapshot found */
    get_history(instr);

    if (!get_tag_ok(instr,SnapShotTag)) {      /* we DO need a SnapShot */
        if (no_mass)
	    warning("There were %d snapshots with no masses. Total mass=1 was set" , no_mass);
	return 0;
    }
        
    get_set(instr, SnapShotTag);
      get_set(instr, ParametersTag);
        get_data(instr, NobjTag, IntType, &nobj, 0);
	if (first) {
	  dprintf(0,"read_snap: Found first snapshot with %d bodies\n",nobj);
	  first = FALSE;
	} 
	dprintf(1,"."); fflush(stderr);
	if (MOBJ == 0) {
	  alloc_snap(nobj);
	  // this will set MOBJ if allowed
	}
	if (nobj>MOBJ)
	  error ("read_snap: not enough memory; only space for %d bodies",MOBJ);
	if ((Qtime=get_tag_ok(instr,TimeTag)))
		get_data(instr,TimeTag,RealType,&tsnap,0);
	else
		tsnap=0.0;
      get_tes(instr,ParametersTag);
      
      if (!get_tag_ok(instr, ParticlesTag)) {		/* do it again, we need ParticlesTag */
	                                                /*   normally happens because of DiagnosticsTag */
	 get_tes(instr,SnapShotTag);			/* close this SnapShotTag */
	 continue;					/* and loop again to read next one */
      }
      dprintf (1,"SnapShot: Time=%f ",tsnap);

      get_set(instr, ParticlesTag);
         get_data(instr, CoordSystemTag, IntType, &cs, 0);
         if (get_tag_ok(instr,MassTag)) {
           get_data_coerced(instr, MassTag, RealType, mass, nobj, 0);
           Qmass=TRUE;
         }
         else if (!Qmass) {
     	      no_mass++;
              dprintf (1,"Warning: no masses present: ASSUME 1/%d\n",nobj);
              for (i=0; i<nobj; i++)
                 mass[i] = 1.0/(double)nobj;
         }
         if (get_tag_ok(instr,PhaseSpaceTag))
	   get_data(instr, PhaseSpaceTag, RealType, &phase[0][0][0], nobj, 2, NDIM, 0);
	 else {
	   get_data_coerced(instr, PosTag, RealType, &phase[0][0][0], nobj, NDIM, 0);
	   get_data_coerced(instr, VelTag, RealType, &phase[0][1][0], nobj, NDIM, 0);
	 }
	 /* need additional Qkey here ? */
	 if (get_tag_ok(instr,KeyTag))
	   get_data(instr, KeyTag, IntType, key, nobj, 0);
	 else {
	   dprintf(1,"added ordinal keys\n");
	   for (i=0; i<nobj; i++) key[i] = i;
	 }
#ifdef ORBIT_PHI
	 get_data_coerced(instr,PotentialTag,RealType,phi,nobj,0);
	 get_data_coerced(instr,AccTag,RealType,&acc[0][0],nobj,NDIM,0);
#endif	 
      get_tes(instr,ParticlesTag);
    get_tes(instr,SnapShotTag);

    return 1;
  }
}

